engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java

```diff
@@ -0,0 +1,508 @@
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+    return selfJavaString.equals(otherJavaString);
+  }
+
+  @Specialization(guards = {
+      "selfInterop.isExecutable(selfFunction)",
+      "otherInterop.isExecutable(otherFunction)",
+      "selfInterop.isMemberInvocable(selfFunction, EQUALS_MEMBER_NAME)",
+      "otherInterop.isMemberInvocable(otherFunction, EQUALS_MEMBER_NAME)",
+  }, limit = "3")
+  boolean equalsFunctions(Object selfFunction, Object otherFunction,
+      @CachedLibrary("selfFunction") InteropLibrary selfInterop,
+      @CachedLibrary("otherFunction") InteropLibrary otherInterop,
+      @CachedLibrary(limit = "3") InteropLibrary retValueInterop) {
+    Object ret;
+    try {
+      ret = selfInterop.invokeMember(selfFunction, MethodNames.Function.EQUALS, otherFunction);
+      return retValueInterop.asBoolean(ret);
+    } catch (UnsupportedMessageException | ArityException | UnknownIdentifierException |
+             UnsupportedTypeException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization(guards = {
+      "selfInterop.hasArrayElements(selfArray)",
+      "otherInterop.hasArrayElements(otherArray)"
+  }, limit = "3")
+  boolean equalsArrays(Object selfArray, Object otherArray,
+      @CachedLibrary("selfArray") InteropLibrary selfInterop,
+      @CachedLibrary("otherArray") InteropLibrary otherInterop,
+      @Cached EqualsNode equalsNode
+      ) {
+    try {
+      long selfSize = selfInterop.getArraySize(selfArray);
+      if (selfSize != otherInterop.getArraySize(otherArray)) {
+        return false;
+      }
+      for (long i = 0; i < selfSize; i++) {
+        Object selfElem = selfInterop.readArrayElement(selfArray, i);
+        Object otherElem = otherInterop.readArrayElement(otherArray, i);
+        if (!equalsNode.execute(selfElem, otherElem)) {
+          return false;
+        }
+      }
+      return true;
+    } catch (UnsupportedMessageException | InvalidArrayIndexException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /**
+   * Behind truffle boundary, because comparing exceptions is not expected to be on fast path.
+   * @return True if the givne exceptions are the same
+   */
+  @TruffleBoundary
+  @Specialization(guards = {
+      "selfInterop.isException(selfException)",
+      "otherInterop.isException(otherException)",
+  }, limit = "3")
+  boolean equalsExceptions(Object selfException, Object otherException,
+      @CachedLibrary("selfException") InteropLibrary selfInterop,
+      @CachedLibrary("otherException") InteropLibrary otherInterop,
+      @CachedLibrary(limit = "5") InteropLibrary interop) {
+    try {
+      boolean sameExceptionTypes =
+          selfInterop.getExceptionType(selfException) == otherInterop.getExceptionType(otherException);
+
+      String selfMessage = null;
+      if (selfInterop.hasExceptionMessage(selfException)) {
```

Comment:
Is this complex behaviour needed? Wouldn't it be simpler to just check for identity? If not, do we have a test for the exception `==` check somewhere?

Reply:
Probably not. It seems like this specialization is not triggered in our tests. It is impossible to do that only in enso, as `(Error.throw "err1") == (Error.throw "err2")` will just return `err1` error. I think that for now, the most sensible thing to do is to remove this specialization altogether, and let these cases be handled by the fallback method.